src/dataservices/highlights.ts

```diff
@@ -70,4 +88,71 @@ export class HighlightsDataService {
     }
     return [];
   }
+
+  /**
+   * Check whether the requested number of highlights would exceed the
+   * highlight limit, taking into account active (status=1) highlights
+   * already associated to the item in the database.
+   * Only necessary to run this check for non-premium users.
+   * @param highlightInput input to create highlight mutation
+   * @throws UserInputError if the requested highlights would exceed
+   * the limit for any item
+   * @returns void if validation passes
+   */
+  async checkHighlightLimit(highlightInput: HighlightInput[]) {
```

Comment:
@wtfluckey here's the tedious check logic -- it was more of a bear than I thought. I pulled out the data reducers we wrote before and made them into utility functions, since I kept reusing them.